### DISCUSSION TITLE
with or map expression to do direct property mapping like AutoMapper does

### DISCUSSION THREAD

**Discussion by avertiy** [dotnet/csharplang#6778] [Language Ideas]
I wondered whether it is possible to add syntax sugar to support direct properties mapping out of the box avoiding AutoMapper  

```
public class Person{
    public string FirstName {get;set;}
    public string LastName {get;set;}
}

public class PersonDTO{
    public string FirstName {get;set;}
    public string LastName {get;set;}
}

var person = new Person() { FirstName ="John", LastName = "Dow"};

var dto= person with {  } as PersonDTO;
```

or
`var dto = person map PersonDTO;`

Boring job to do mapping manually but using AutoMapper sometimes is more headache then the gain of it.
It would be nice if C# would support simple properties mapping natively (no need to cover all AutoMapper mapping cases,  this is for simple direct mapping case when AutoMapper seems too big gun to use it.

**Comment by svick** [comment]
> using AutoMapper sometimes is more headache then the gain of it, so C# support this natively will make life easier

What kind of headaches would this avoid?

**Reply by avertiy** [reply]
here is an example and many many entities mapped like this without AutoMapper 
I know this is a job for AutoMapper but i don't want to do extra things like creating/managing  mapper profiles, extra dependencies etc.

no need in AutoMapper special features like null handling, resolving props, simply direct mapping

`var clientDto = clients.Select(x => new ClientDTO
            {
                Id = x.Id,
                Email = x.Email,
                IsEmailVerified = x.IsEmailVerified,
                CurrencyId = x.CurrencyId,
                UserName = x.UserName,
                PartnerId = x.PartnerId,
                Gender = x.Gender,
                BirthDate = x.BirthDate,
                State = x.State,
                CategoryId = x.CategoryId,
                FirstName = x.FirstName,
                LastName = x.LastName,
                NickName = x.NickName,
                SecondName = x.SecondName,
                SecondSurname = x.SecondSurname,
                ...
            }).FirstOrDefaultAsync();`

**Reply by avertiy** [reply]
+manual mapping is error prone when new property is added, the mapping might miss it and many projects or small github repos where AutoMapper would be a big gun for a small bird..

**Reply by spydacarnage** [reply]
If you use the new `required` keyword on your destination class properties, the compiler will warn you if you forget to copy a new property.

**Comment by HaloFour** [comment]
Automappers feel like an excellent use case for source generators.  As a language feature it would have to be intentionally quite limited which would create a huge cliff as soon as you hit any kind of difference in the schemas between the types, but source generators can control that via policies that could be tweaked in a myriad of different ways, including custom attributes.

**Reply by spydacarnage** [reply]
Having just about finished a SG-based mapper for my own project, it's a wonder to use (but was a PITA to get right, especially nullability and implicit/explicit casting!)

But it's certainly not something I could even imagine having built in to the language - it would need too much user preference to work.

**Comment by TahirAhmadov** [comment]
Specifically because mapping properties can get complicated, it should NOT be in the language. What about conversions? Sub-objects? Collections? Handling of nulls? The language team would essentially endeavor to implement a mapping tool, but without the benefit of making it possible to configure, because a language has to work for everybody.

**Reply by avertiy** [reply]
Ok the idea is not to replace AutoMapper at all with all its features, but avoid the need of  AutoMapper for simple case where only direct properties mapping take place like in the example above

**Reply by TahirAhmadov** [reply]
So now, you have 2 ways of doing the same thing? And while you are happy with the simplest scenario, what about others who will undoubtedly start requesting more features? Another problem with this is it erodes the notion of type safety, when it's a language feature like that, to "shape-map" types to each other.

**Reply by TahirAhmadov** [reply]
Also, the simplest situations are also easily handled by AutoMapper, so there is no need to add a language feature for something which is trivial using an existing library.

**Reply by avertiy** [reply]
_others who will undoubtedly start requesting more features_ 
ahaha, don't worry, C# already overcomplicated, 50% of its features is like a junk constructs which people don't use, rarely use or avoiding to use, one day C# will be reworked completely like .NET Core replaced .NET Framework.. the time will vote what's essential and what's just a junk in the language and could be dropped out to make the language clean.

**Reply by HaloFour** [reply]
> 50% of its features is like a junk constructs which people don't use

Except that everyone has a completely different 50%.  The language team would vehemently disagree with you that most, or really any, of the language is junk features that have no use or that nobody uses.

**Reply by avertiy** [reply]
that's another topic to discuss, I expect in 5 years or so the new language will appear, C => C++  => C#. The C# is 20 y.o and its era comes to the end. When I was a student i didn't want to learn neither C++ neither Java, C# world seemed newer  and easier than overcomplicated C++ world.  Why 20 years old student today would be eager to learn old C#? They learn js React etc. Once Microsoft will realize that new generation passes by they will bring the C##  or G#..

**Reply by HaloFour** [reply]
> that's another topic to discuss, I expect in 5 years or so the new language will appear

I expect that in 5 years the TIOBE index will look much like it does today, dominated by 25+ year old languages like C#, C++, Java and Python.  The age of the language doesn't matter very much.

> Why 20 years old student today would be eager to learn old C#?

The 20 year old student should learn a variety of languages and paradigms with a multitude of tools.  But which particular languages they learn isn't that relevant given the skills are transferable.  Someone who understands Javascript (which is older and substantially more crufty) should be able to pick up C# without too much trouble.  Ultimately, what they use outside of their own hobby projects will be dictated by the opportunities that they find in the market.  Afterall, I seriously doubt anyone uses a mapper because it's fun.
